### PR TITLE
[4.0] Fix debug logging. Move logging to application

### PR DESCRIPTION
**Pull request by Fedik**
This is try to fix  Release Blocker #25675 .

### Summary of Changes

This will move logging set up to application level. The plugin used only for display them.
One of new feature here, that the logging may be enabled even when the debugging is Off.

### Testing Instructions

Be aware: there a thousands of Deprecated entries, if you have a slow PC / server, be patient.

Apply patch. Go to Global Configuration and enable Debug, also enable Log Deprecated API.
Enable System Debug plugin, and set "Show logs" in the plugin.

In the debugger toolbar you should be able to see much of entries of Deprecated call.

### Actual result BEFORE applying this Pull Request

No Deprecated entries, or only few of them.

### Expected result AFTER applying this Pull Request

A loot of entries

### Documentation Changes Required

Yes I guess. Logging moved to Global configuration, and can be enabled there.

### CODE REVIEW

File: administrator/components/com_config/forms/application.xml [hunk 1]
```diff
@@ -1098,18 +1098,89 @@
...
-			label="COM_CONFIG_FIELD_LOG_PATH_LABEL"
-			required="true"
-			filter="string"
-			size="50"
+				name="log_path"
+				type="text"
+				label="COM_CONFIG_FIELD_LOG_PATH_LABEL"
+				required="true"
+				filter="string"
+				size="50"
```
Comment (brianteeman) [inline administrator/components/com_config/forms/application.xml]:
```suggestion
			name="log_path"
			type="text"
			label="COM_CONFIG_FIELD_LOG_PATH_LABEL"
			required="true"
			filter="string"
			size="50"
```

File: administrator/components/com_config/forms/application.xml [hunk 2]
```diff
@@ -1098,18 +1098,89 @@
...
+		/>
+
+		<field
+				name="log_everything"
+				type="radio"
+				label="COM_CONFIG_FIELD_LOG_EVERYTHING_LABEL"
+				description="COM_CONFIG_FIELD_LOG_EVERYTHING_DESC"
+				layout="joomla.form.field.radio.switcher"
+				default="0"
+				filter="integer"
```
Comment (brianteeman) [inline administrator/components/com_config/forms/application.xml]:
```suggestion
			name="log_everything"
			type="radio"
			label="COM_CONFIG_FIELD_LOG_EVERYTHING_LABEL"
			description="COM_CONFIG_FIELD_LOG_EVERYTHING_DESC"
			layout="joomla.form.field.radio.switcher"
			default="0"
			filter="integer"
```

File: administrator/components/com_config/forms/application.xml [hunk 3]
```diff
@@ -1098,18 +1098,89 @@
...
+		>
```
Comment (brianteeman) [inline administrator/components/com_config/forms/application.xml]:
```suggestion
			>
```

File: administrator/components/com_config/forms/application.xml [hunk 4]
```diff
@@ -1098,18 +1098,89 @@
...
+		</field>
+
+		<field
+				name="log_deprecated"
+				type="radio"
+				label="COM_CONFIG_FIELD_LOG_DEPRECATED_LABEL"
+				layout="joomla.form.field.radio.switcher"
+				default="0"
+				filter="integer"
+		>
```
Comment (brianteeman) [inline administrator/components/com_config/forms/application.xml]:
```suggestion
			name="log_deprecated"
			type="radio"
			label="COM_CONFIG_FIELD_LOG_DEPRECATED_LABEL"
			layout="joomla.form.field.radio.switcher"
			default="0"
			filter="integer"
			>
```

File: administrator/components/com_config/forms/application.xml [hunk 5]
```diff
@@ -1098,18 +1098,89 @@
...
+
+		<field
+				name="log_priorities"
+				type="list"
+				label="COM_CONFIG_FIELD_LOG_PRIORITIES_LABEL"
+				multiple="true"
+				layout="joomla.form.field.list-fancy-select"
+				default="all"
+				validate="options"
+		>
```
Comment (brianteeman) [inline administrator/components/com_config/forms/application.xml]:
```suggestion
			name="log_priorities"
			type="list"
			label="COM_CONFIG_FIELD_LOG_PRIORITIES_LABEL"
			multiple="true"
			layout="joomla.form.field.list-fancy-select"
			default="all"
			validate="options"
			>
```

File: administrator/components/com_config/forms/application.xml [hunk 6]
```diff
@@ -1098,18 +1098,89 @@
...
+			<option value="debug">COM_CONFIG_FIELD_LOG_PRIORITIES_DEBUG</option>
+		</field>
+
+		<field
+				name="log_categories"
+				type="text"
+				label="COM_CONFIG_FIELD_LOG_CATEGORIES_LABEL"
+				description="COM_CONFIG_FIELD_LOG_CATEGORIES_DESC"
+				size="60"
+				default=""
```
Comment (brianteeman) [inline administrator/components/com_config/forms/application.xml]:
```suggestion
			name="log_categories"
			type="text"
			label="COM_CONFIG_FIELD_LOG_CATEGORIES_LABEL"
			description="COM_CONFIG_FIELD_LOG_CATEGORIES_DESC"
			size="60"
			default=""
```

File: administrator/components/com_config/forms/application.xml [hunk 7]
```diff
@@ -1098,18 +1098,89 @@
...
 		/>
 
+		<field
+				name="log_category_mode"
+				type="radio"
+				label="COM_CONFIG_FIELD_LOG_CATEGORY_MODE_LABEL"
+				layout="joomla.form.field.radio.switcher"
+				default="0"
+				filter="integer"
+		>
```
Comment (brianteeman) [inline administrator/components/com_config/forms/application.xml]:
```suggestion
			name="log_category_mode"
			type="radio"
			label="COM_CONFIG_FIELD_LOG_CATEGORY_MODE_LABEL"
			layout="joomla.form.field.radio.switcher"
			default="0"
			filter="integer"
			>
```

File: libraries/src/Log/Logger/InMemoryLogger.php [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+
+namespace Joomla\CMS\Log\Logger;
+
+\defined('JPATH_PLATFORM') or die;
+
+use Joomla\CMS\Log\LogEntry;
+use Joomla\CMS\Log\Logger;
+
+/**
+ * Logger class that keep all entries in memory
```
Comment (brianteeman) [inline libraries/src/Log/Logger/InMemoryLogger.php]:
```suggestion
 * Logger class that keeps all entries in memory
```

**Comment by toivo** [comment]
I have tested this item :white_check_mark: successfully on 0275ab9e9b6ae8a2d038686d9824ca894776e031<br /><br />Tested successfully in Beta4-dev of 5 August on Wampserver 3.2.2 using PHP 7.4.8.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30317">issues.joomla.org/tracker/joomla-cms/30317</a>.</sub>

File: administrator/language/en-GB/com_config.ini [hunk 9]
```diff
@@ -84,7 +84,25 @@ COM_CONFIG_FIELD_FTP_ROOT_LABEL="FTP Root"
 COM_CONFIG_FIELD_FTP_USERNAME_LABEL="FTP Username"
 COM_CONFIG_FIELD_GZIP_COMPRESSION_LABEL="Gzip Page Compression"
 COM_CONFIG_FIELD_HTMLBODY_LABEL="HTML Body"
+COM_CONFIG_FIELD_LOG_CATEGORIES_DESC="A comma separated list of log categories to include. Common log categories include but are not limited to: database, databasequery, database-error, deprecated and jerror. If empty, all categories will be shown."
+COM_CONFIG_FIELD_LOG_CATEGORIES_LABEL="Log Categories"
+COM_CONFIG_FIELD_LOG_CATEGORY_MODE_EXCLUDE="Exclude"
+COM_CONFIG_FIELD_LOG_CATEGORY_MODE_INCLUDE="Include"
+COM_CONFIG_FIELD_LOG_CATEGORY_MODE_LABEL="Log Category Mode"
+COM_CONFIG_FIELD_LOG_DEPRECATED_LABEL="Log Deprecated API"
+COM_CONFIG_FIELD_LOG_EVERYTHING_LABEL="Log Almost Everythin"
```
Comment (richard67) [inline administrator/language/en-GB/com_config.ini]:
```suggestion
COM_CONFIG_FIELD_LOG_EVERYTHING_LABEL="Log Almost Everything"
```

**Review by richard67 — CHANGES_REQUESTED** [review]

**Review by richard67 — APPROVED** [review]

**Comment by richard67** [comment]
Previous test by @toivo is still valid because last change only was a typo correction in a language string. I've restored the test result in the issue tracker.

**Comment by richard67** [comment]
I have tested this item :white_check_mark: successfully on 31e7e7764bb1489a0fd5b52aa53d6836d096ae4e<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30317">issues.joomla.org/tracker/joomla-cms/30317</a>.</sub>

**Comment by richard67** [comment]
@Fedik To save you time, I've allowed myself to commit the typo correction for the language string directly to your branch. I hope you are ok with that. If not, let me know and I'll never do it again.

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30317">issues.joomla.org/tracker/joomla-cms/30317</a>.</sub>

**Comment by ohrionmartin** [comment]
I have tested this item :white_check_mark: successfully on 31e7e7764bb1489a0fd5b52aa53d6836d096ae4e<br /><br />System Information
=============
php: Linux lamp121.cloudaccess.net 2.6.32-673.8.1.lve1.4.3.1.el6.x86_64 #1 SMP Thu May 5 16:06:47 EDT 2016 x86_64
dbserver: mysql
dbversion: 5.7.29-cll-lve
dbcollation: utf8_general_ci
dbconnectioncollation: utf8mb4_general_ci
dbconnectionencryption: 
dbconnencryptsupported: true
phpversion: 7.3.19
server: Apache
sapi_name: cgi-fcgi
version: Joomla! 4.0.0-beta4-dev Development [ Mañana ] 29-July-2020 18:21 GMT
useragent: Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:79.0) Gecko/20100101 Firefox/79.0
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/30317">issues.joomla.org/tracker/joomla-cms/30317</a>.</sub>

**Comment by richard67** [comment]
@wilsonge Anything keeps this from being merged?

File: libraries/src/Application/CMSApplication.php [hunk 10]
```diff
@@ -1309,4 +1311,58 @@ private function hasUserConfiguredTwoFactorAuthentication(): bool
 
 		return true;
 	}
+
+	/**
+	 * Setup logging functionality.
+	 *
+	 * @return void
+	 *
+	 * @since 4.0.0
```
Comment (zero-24) [inline libraries/src/Application/CMSApplication.php]:
```suggestion
	 * @since   __DEPLOY_VERSION__
```

**Pull request merged**